go/internal/cbor/deterministic.go

```diff
@@ -0,0 +1,121 @@
+package cbor
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+)
+
+// Deterministic function loops through the given byte array containing CBOR encoded
+// byte sequences and checks that they follow the deterministic principles described here:
+// https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c.
+func Deterministic(input []byte) bool {
+	index := 0
+
+	for index < len(input) {
+		// TODO(sonkkeli): Remove debug print.
+		fmt.Printf("index=%v\n", index)
+
+		length, err := deterministicRec(input[index:])
+		if err != nil {
+			fmt.Println(err)
+			return false
+		}
+		index += length + 1
+	}
+	return true
+}
+
+// deterministicRec is a recursively called helper function to check the deterministicy of a bytes array.
+// TODO(sonkkeli): Recursively called parts will be the maps and arrays, which contain other types, meaning
+// unsigned integers, byte strings, text strings.
+func deterministicRec(input []byte) (int, error) {
+	cborItemLengthInBytes := 0
+
+	b := input[0]
+
+	switch getMajorType(b) {
+	case TypePosInt:
+		length, err := unsignedIntegerDeterministic(input)
+		if err != nil {
+			return 0, err
+		}
+		cborItemLengthInBytes = length
+
+		break
+
+	case TypeBytes, TypeText:
+		// TODO(sonkkeli):
+		// length, err := textOrBytesDeterministic(input[index:])
+		break
+
+	case TypeArray:
+		// TODO(sonkkeli):
+		// length, err := arrayDeterministic(input[index:])
+		break
+
+	case TypeMap:
+		// TODO(sonkkeli):
+		// length, err := mapDeterministic(input[index:])
+		break
+
+	default:
+		return 0, errors.New("Missing implementation for major type.")
+	}
+
+	// TODO(sonkkeli): Remove debug print.
+	fmt.Printf("cborItemLengthInBytes=%v\n", cborItemLengthInBytes)
+
+	return cborItemLengthInBytes, nil
+
+}
+
+// unsignedIntegerDeterministic calculates the value of the unsigned integer to ensure that
+// the right amount of bytes is used in the CBOR encoding and returns the length of the unsigned integer.
+func unsignedIntegerDeterministic(input []byte) (int, error) {
+	ainfo := convertToAdditionalInfo(input[0])
+
+	if ainfo == AdditionalInfoIndefinite || ainfo == AdditionalInfoReserved {
+		return -1, fmt.Errorf("%v should not be in use in deterministic CBOR.", ainfo)
+	}
+
+	lengthInBytes, err := ainfo.getAdditionalInfoLength()
+	if err != nil {
+		return -1, err
+	}
+
+	_, err = unsignedIntegerCompareValueToLimit(input[0 : lengthInBytes+1])
+	if err != nil {
+		return -1, err
+	}
+
+	return lengthInBytes, nil
```

Comment:
Shouldn't be `lengthInBytes + 1`?

Reply:
The +1 is in the main loop :)

Reply:
It's not length -1, it is the right length and then the mainloop is where it keeps track of the CBOR sequences. So if there's like e.g. 2 unsigned integers behind each other of length 2 and 3, the length of the 1st one is calculated and deterministicRec returns 2, then the main function Deterministic notices that there's still something left on the byte array and it increases the index on the main loop to the first byte of the second CBOR item (of len 3) in the sequence and then the deterministicRec returns 3 for that. Then on the main loop it notices that index is no longer less than length of input and the function exits. I am actually not sure if it makes any difference to move the +1 inside deterministicRec for when arrays and maps come, maybe not? I have to code the rest and test

Reply:
I'm not sure if I'm following you. If the first `deterministicRec` call returns 2, the second `deterministicRec` is called with `input[3:]`. But the second unsigned integer is located at `index[2:5]`, right?

```
 [0] [1] [2] [3] [4]
+---+---+---+---+---+
| first |  second   |
+---+---+---+---+---+
```